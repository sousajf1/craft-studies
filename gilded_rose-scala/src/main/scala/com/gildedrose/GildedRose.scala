package com.gildedrose

import com.gildedrose.GildedRose.{AGEDBRIE, BACKSTAGE, SULFURAS}


class GildedRose(val items: Array[Item]) {
  def updateQuality() {
    items.foreach { item =>
      item.name match {
        case AGEDBRIE =>
          updateAgedBrie(item)
        case BACKSTAGE =>
          updateBackstage(item)
        case SULFURAS =>
          updateSulfuras(item)
        case _ =>
          updateDefault(item)
      }
    }
  }

  private def updateDefault(item: Item) = {
    if (item.quality > 0) {
      item.quality = item.quality - 1
    }

    item.sellIn = item.sellIn - 1

    if (item.sellIn < 0) {
      if (item.quality > 0) {
        item.quality = item.quality - 1
      }
    }
  }

  private def updateSulfuras(item: Item) = {

  }

  private def updateBackstage(item: Item) = {
    if (item.quality < 50) {
      item.quality = item.quality + 1
      if (item.sellIn < 11) {
        if (item.quality < 50) {
          item.quality = item.quality + 1
        }
      }
      if (item.sellIn < 6) {
        if (item.quality < 50) {
          item.quality = item.quality + 1
        }
      }
    }
    item.sellIn = item.sellIn - 1
    if (item.sellIn < 0) {
      item.quality = 0
    }
  }

  private def updateAgedBrie(item: Item) {
    if (item.quality < 50) {
      item.quality = item.quality + 1
    }
    item.sellIn = item.sellIn - 1
    if (item.sellIn < 0) {
      if (item.quality < 50) {
        item.quality = item.quality + 1
      }
    }
  }

}

object GildedRose {
  private val AGEDBRIE = "Aged Brie"
  private val SULFURAS = "Sulfuras, Hand of Ragnaros"
  private val BACKSTAGE = "Backstage passes to a TAFKAL80ETC concert"
}