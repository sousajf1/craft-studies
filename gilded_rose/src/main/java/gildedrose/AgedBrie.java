package gildedrose;

public class AgedBrie extends ItemUpdatable {
    public AgedBrie(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }


}
