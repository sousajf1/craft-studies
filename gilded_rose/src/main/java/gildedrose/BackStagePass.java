package gildedrose;

public class BackStagePass extends ItemUpdatable {
    public BackStagePass(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }


}
