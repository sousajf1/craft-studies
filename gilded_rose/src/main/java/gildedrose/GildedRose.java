package gildedrose;
import java.util.ArrayList;
import java.util.List;


public class GildedRose {

	public static List<Item> items = null;

    public GildedRose(List<Item> items) {
        this.items = items;
    }

    /**
	 * @param args
	 */
	public static void main(String[] args) {
		
        System.out.println("OMGHAI!");
		
        items = new ArrayList<Item>();
        items.add(new Item("+5 Dexterity Vest", 10, 20));
        items.add(new Item("Aged Brie", 2, 0));
        items.add(new Item("Elixir of the Mongoose", 5, 7));
        items.add(new Item("Sulfuras, Hand of Ragnaros", 0, 80));
        items.add(new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20));
        items.add(new Item("Conjured Mana Cake", 3, 6));

        updateQuality();
}


	
    public static void updateQuality()
    {
        for (int i = 0; i < items.size(); i++)
        {
            doUpdateQuality(items.get(i));
        }
    }

    private static void doUpdateQuality(Item item) {
        switch (item.getName()) {
            case "Aged Brie":
                if (item.getQuality() < 50) {
                    item.setQuality(item.getQuality() + 1);

                }

                item.setSellIn(item.getSellIn() - 1);

                if (item.getSellIn() < 0) {
                    if (item.getQuality() < 50) {
                        item.setQuality(item.getQuality() + 1);
                    }
                }
                break;
            case "Backstage passes to a TAFKAL80ETC concert":
                if (item.getQuality() < 50) {
                    item.setQuality(item.getQuality() + 1);

                    if (item.getSellIn() < 11) {
                        if (item.getQuality() < 50) {
                            item.setQuality(item.getQuality() + 1);
                        }
                    }

                    if (item.getSellIn() < 6) {
                        if (item.getQuality() < 50) {
                            item.setQuality(item.getQuality() + 1);
                        }
                    }
                }

                item.setSellIn(item.getSellIn() - 1);

                if (item.getSellIn() < 0) {
                    item.setQuality(0);
                }
                break;
            case "Sulfuras, Hand of Ragnaros":

                break;
            default:
                if (item.getQuality() > 0) {
                    item.setQuality(item.getQuality() - 1);
                }

                item.setSellIn(item.getSellIn() - 1);

                if (item.getSellIn() < 0) {
                    if (item.getQuality() > 0) {
                        item.setQuality(item.getQuality() - 1);
                    }
                }
                break;
        }
    }

}