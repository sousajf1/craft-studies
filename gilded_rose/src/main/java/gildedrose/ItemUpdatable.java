package gildedrose;

public class ItemUpdatable extends Item {
    ItemUpdatable(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }

    public static ItemUpdatable createItemUpdatable(String name, int sellIn, int quality) {
        switch (name) {
            case "Aged Brie":
                return new AgedBrie(name, sellIn, quality);
            case "Backstage passes to a TAFKAL80ETC concert":
                return new BackStagePass(name, sellIn, quality);
            case "Sulfuras, Hand of Ragnaros":
                return new SulfurasHand(name, sellIn, quality);
            default:
                return new ItemUpdatable(name, sellIn, quality);
        }
    }

    protected void doUpdateQuality() {
        if (getQuality() > 0) {
            setQuality(getQuality() - 1);
        }

        setSellIn(getSellIn() - 1);

        if (getSellIn() < 0) {
            if (getQuality() > 0) {
                setQuality(getQuality() - 1);
            }
        }
    }
}

