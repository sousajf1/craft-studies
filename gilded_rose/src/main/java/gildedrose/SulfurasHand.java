package gildedrose;

public class SulfurasHand extends ItemUpdatable {
    public SulfurasHand(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }

    @Override
    protected void doUpdateQuality() {
    }
}
