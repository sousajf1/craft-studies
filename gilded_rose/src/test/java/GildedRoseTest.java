import gildedrose.GildedRose;
import gildedrose.Item;
import org.approvaltests.Approvals;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.Test;

import java.util.*;


public class GildedRoseTest {

	@Test 
	public void update_quality_approval_test() {
		String name = "newItem";
		int sellin = 0;
		int quality = 0;
		String itemString = doUpdate(name, sellin, quality);
		//Approvals.verify(itemString);
		CombinationApprovals.verifyAllCombinations(this::doUpdate,
													new String[] {name},
													new Integer[] {sellin},
													new Integer[] {quality});
	}

	private String doUpdate(String name, int sellin, int quality) {
		List<Item> items = new ArrayList<Item>();
		items.add(new Item(name,sellin,quality));
		GildedRose app = new GildedRose(items);
		app.updateQuality();
		return app.items.get(0).name + " , " +
				app.items.get(0).sellIn + " , " +
				app.items.get(0).quality;
	}
}
