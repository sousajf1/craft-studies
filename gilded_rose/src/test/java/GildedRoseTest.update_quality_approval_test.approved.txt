[newItem, 0, 0] => newItem , -1 , 0 
